File: solutions/Two Pointers/Medium/11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python3
#
# [11] Container With Most Water
#

'''
condition: List[int] height, length = n
goal: get the max size
solution: 2ptrs
'''

# @lc code=start
class Solution:
    def maxArea(self, height: List[int]) -> int:
        max_size = 0

        left, right = 0, len(height)-1

        while left < right:

            current_height = min(height[left], height[right])

            size = (right-left) * current_height

            max_size = max(max_size, size)

            if height[left] < height[right]:
                left += 1
            else: right -= 1

        return max_size


        
# @lc code=end

File: solutions/Two Pointers/Medium/11.container-with-most-water.java
/*
 * @lc app=leetcode id=11 lang=java
 *
 * [11] Container With Most Water
 * DP: 原理就是，出现短的那一头，就移动一下，看看下一根是否会高。
 */

// @lc code=start
class Solution {
    public int maxArea(int[] height) {

        int left = 0;
        int right = height.length-1;
        int max_area = 0;

        while (left < right) {

            int length = right - left;
            int width = Math.min(height[left], height[right]);

            int area = length * width;

            if (height[left] > height[right]) right--;
            else left++;

            max_area = Math.max(max_area, area);
        }

        return max_area;
     
        
    }
}
// @lc code=end

